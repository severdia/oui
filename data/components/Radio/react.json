{"description":"Generates a `input` element of type `radio` that is wrapped in a `Label`.","methods":[],"props":{"checked":{"type":{"name":"bool"},"required":false,"description":"Boolean to set radio input, for controlled component"},"defaultChecked":{"type":{"name":"bool"},"required":false,"description":"Boolean for how radio input renders initially"},"isDisabled":{"type":{"name":"bool"},"required":false,"description":"Prevents radio input from being modified and appears disabled"},"label":{"type":{"name":"string"},"required":false,"description":"Text that describes the radio input"},"name":{"type":{"name":"string"},"required":true,"description":"String that can be used to identify a set of radio inputs so that only one\n    in the set is checked at any given time."},"onChange":{"type":{"name":"func"},"required":false,"description":"Function that fires when the radio input is clicked"},"testSection":{"type":{"name":"string"},"required":false,"description":"Hook for automated JavaScript tests","defaultValue":{"value":"null","computed":false}}},"private":false}