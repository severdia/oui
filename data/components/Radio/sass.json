{"name":"Radio\n","description":"","start":{"line":1,"column":1},"end":{"line":14,"column":4},"example":[{"type":"html","description":null,"code":"<ul class=\"oui-input-list\">\n  <li>\n    <input name=\"radio-set-1\" id=\"bar-1\" type=\"radio\" checked>\n    <label class=\"oui-label\" for=\"bar-1\">Radio Option #1\n  </li>\n  <li>\n    <input name=\"radio-set-1\" id=\"bar-2\" type=\"radio\">\n    <label class=\"oui-label\" for=\"bar-2\">Radio Option #2\n  </li>\n</ul>"}]}