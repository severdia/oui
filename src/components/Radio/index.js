import React from 'react';
import PropTypes from 'prop-types';

import classNames from 'classnames';
import Label from '../Label';

/**
 * Generates a `input` element of type `radio` that is wrapped in a `Label`.
 * @param {Object} props - Properties passed to component
 * @returns {ReactElement}
 */
const Radio = ({
  isDisabled,
  testSection = null,
  name,
  defaultChecked,
  checked,
  onChange,
  label,
  labelWeight,
}) => {
  const labelClassNames = classNames({
    'flush--bottom': true,
    'push--left': true,
    'weight--light': labelWeight === 'light',
    'weight--normal': labelWeight === 'normal',
    'weight--bold': labelWeight === 'bold',
    'cursor--pointer': true,
    'oui-label--disabled': isDisabled,
  });

  const classes = classNames({
    'flex--none': true,
  });

  return (
    <Label data-test-section={ `${testSection}-label` }>
      <div className="flex">
        <input
          data-oui-component={ true }
          type="radio"
          name={ name }
          defaultChecked={ defaultChecked }
          checked={ checked }
          className={ classes }
          disabled={ isDisabled }
          onChange={ onChange }
          data-test-section={ testSection }
          style={{ 'marginTop': '0.3em' }}
        />
        <div className={ labelClassNames }>
          { label }
        </div>
      </div>
    </Label>
  );
};

Radio.propTypes = {
  /** Boolean to set radio input, for controlled component  */
  checked: PropTypes.bool,
  /** Boolean for how radio input renders initially  */
  defaultChecked: PropTypes.bool,
  /** Prevents radio input from being modified and appears disabled */
  isDisabled: PropTypes.bool,
  /** Text that describes the radio input */
  label: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  /** Font weight for the label text */
  labelWeight: PropTypes.oneOf(['light', 'normal', 'bold']),
  /**
    String that can be used to identify a set of radio inputs so that only one
    in the set is checked at any given time.
  */
  name: PropTypes.string.isRequired,
  /** Function that fires when the radio input is clicked */
  onChange: PropTypes.func,
  /** Hook for automated JavaScript tests */
  testSection: PropTypes.string,
};

Radio.defaultProps = {
  labelWeight: 'normal',
};

export default Radio;
