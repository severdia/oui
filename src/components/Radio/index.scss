////
/// Radio
/// @example[html]
///   <ul class="oui-input-list">
///     <li>
///       <input name="radio-set-1" id="bar-1" type="radio" checked>
///       <label class="oui-label" for="bar-1">Radio Option #1
///     </li>
///     <li>
///       <input name="radio-set-1" id="bar-2" type="radio">
///       <label class="oui-label" for="bar-2">Radio Option #2
///     </li>
///   </ul>
////


// This file is only needed to display the examples.
