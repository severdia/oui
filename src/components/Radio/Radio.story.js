import React from 'react';

import { storiesOf } from '@storybook/react';
import { withKnobs, text, select } from '@storybook/addon-knobs';

import Radio from './index.js';

const stories = storiesOf('Radio', module);
stories
  .addDecorator(withKnobs)
  .addDecorator(story => (
    <div id="root-preview">
      {story()}
    </div>
  ));

stories
  .add('Default', (() => {
    return (
      <Radio
        testSection={ text('testSection', 'test-radio') }
        name={ text('name', 'this is a radio') }
        label={ text('label', 'this is a radio') }
      />
    );
  }))
  .add('With html label', (() => {
    return (
      <Radio
        testSection={ text('testSection', 'test-radio') }
        name={ text('name', 'this is a radio') }
        label={
          <div>
            <strong>Label</strong>
            <br />
            <span>Description text</span>
          </div>
        }
      />
    );
  }))
  .add('With labelWeight prop', (() => {
    return (
      <Radio
        testSection={ text('testSection', 'test-radio') }
        name={ text('name', 'this is a radio') }
        label={ text('label', 'this is a radio') }
        labelWeight={ select('labelWeight', ['light', 'normal', 'bold'], 'bold') }
      />
    );
  }));
