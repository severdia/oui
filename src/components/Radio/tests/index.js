import React from 'react';
import Radio from '../index';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

describe('components/Radio', () => {
  const mockFunction = jest.fn();

  it('should render an input of type radio', () => {
    const component = shallow(<Radio name="goose" />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should render input with name when provided', () => {
    const component = shallow(<Radio name="goose" />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should render a checked radio input when `defaultChecked` is true', () => {
    const component = shallow(<Radio name="goose" defaultChecked={ true } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should render an unchecked radio input when `defaultChecked` is false', () => {
    const component = shallow(<Radio name="goose" defaultChecked={ false } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should render a checked radio input when `checked` is true', () => {
    const component = shallow(<Radio name="goose" checked={ true } onChange={ mockFunction } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should render an unchecked radio input when `checked` is false', () => {
    const component = shallow(<Radio name="goose" checked={ false } onChange={ mockFunction } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should throw an error if a radio input is rendered without an `onChange` function', () => {
    spyOn(console, 'error').and.stub();

    mount(<Radio name="goose" checked={ true } />);

    expect(console.error.calls.all()[0].args[0]).toContain('You provided a `checked` prop to a form field without an `onChange` handler'); // eslint-disable-line
  });

  it('should call the `onChange` event handler when the radio input is changed', () => {
    const handler = {
      onChange: (event) => {},
    };
    spyOn(handler, 'onChange');

    const component = shallow(<Radio name="goose" onChange={ handler.onChange } />);

    component.find('input').simulate('change');
    expect(handler.onChange).toHaveBeenCalled();
  });

  it('should be wrapped in a label', () => {
    const component = shallow(<Radio name="goose" />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('label should render text', () => {
    const component = shallow(<Radio name="goose" label="foo" />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should add disabled class to label text when disabled', () => {
    const component = shallow(<Radio name="goose" isDisabled={ true } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should not add disabled class to label text when not disabled', () => {
    const component = shallow(<Radio name="goose" isDisabled={ false } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should add disabled attribute to radio input when disabled', () => {
    const component = shallow(<Radio name="goose" isDisabled={ true } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should not add disabled attribute to radio input when not disabled', () => {
    const component = shallow(<Radio name="goose" isDisabled={ false } />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should add weight--normal class to label text when no labelWeight prop present', () => {
    const component = mount(<Radio />);

    const label = component.find('.weight--normal');
    expect(label.length).toBe(1);
  });

  it('should add weight--normal class to label text when labelWeight prop = "normal"', () => {
    const component = mount(<Radio labelWeight="normal" />);

    const label = component.find('.weight--normal');
    expect(label.length).toBe(1);
  });

  it('should add weight--light class to label text when labelWeight prop = "light"', () => {
    const component = mount(<Radio labelWeight="light" />);

    const label = component.find('.weight--light');
    expect(label.length).toBe(1);
  });

  it('should add weight--bold class to label text when labelWeight prop = "bold"', () => {
    const component = mount(<Radio labelWeight="bold" />);

    const label = component.find('.weight--bold');
    expect(label.length).toBe(1);
  });

  it('should ====OJO====', () => {
    const radioChange = jest.fn();
    const radio = mount(<Radio onChange={ radioChange } name="goose" />);
    radio.find('input').simulate('change');
    expect(radioChange).toHaveBeenCalled();
  });
});
