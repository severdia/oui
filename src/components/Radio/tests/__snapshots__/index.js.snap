// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Radio label should render text 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      name="goose"
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    >
      foo
    </div>
  </div>
</Label>
`;

exports[`components/Radio should add disabled attribute to radio input when disabled 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      disabled={true}
      name="goose"
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer oui-label--disabled"
    />
  </div>
</Label>
`;

exports[`components/Radio should add disabled class to label text when disabled 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      disabled={true}
      name="goose"
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer oui-label--disabled"
    />
  </div>
</Label>
`;

exports[`components/Radio should be wrapped in a label 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      name="goose"
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;

exports[`components/Radio should not add disabled attribute to radio input when not disabled 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      disabled={false}
      name="goose"
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;

exports[`components/Radio should not add disabled class to label text when not disabled 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      disabled={false}
      name="goose"
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;

exports[`components/Radio should render a checked radio input when \`checked\` is true 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      checked={true}
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      name="goose"
      onChange={[MockFunction]}
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;

exports[`components/Radio should render a checked radio input when \`defaultChecked\` is true 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      defaultChecked={true}
      name="goose"
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;

exports[`components/Radio should render an input of type radio 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      name="goose"
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;

exports[`components/Radio should render an unchecked radio input when \`checked\` is false 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      checked={false}
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      name="goose"
      onChange={[MockFunction]}
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;

exports[`components/Radio should render an unchecked radio input when \`defaultChecked\` is false 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      defaultChecked={false}
      name="goose"
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;

exports[`components/Radio should render input with name when provided 1`] = `
<Label
  data-test-section="null-label"
>
  <div
    className="flex"
  >
    <input
      className="flex--none"
      data-oui-component={true}
      data-test-section={null}
      name="goose"
      style={
        Object {
          "marginTop": "0.3em",
        }
      }
      type="radio"
    />
    <div
      className="flush--bottom push--left weight--normal cursor--pointer"
    />
  </div>
</Label>
`;
